skimage/measure/profile.py

```diff
@@ -68,14 +80,10 @@ def profile_line(image, src, dst, linewidth=1,
         pixels = ndi.map_coordinates(image, perp_lines,
                                      order=order, mode=mode, cval=cval)
 
-    if intensity_type == 'mean':
-        intensities = pixels.mean(axis=1)
-    elif intensity_type == 'max' or 'maximum':
-        intensities = pixels.max(axis=1)
-    elif intensity_type == 'min' or 'minimum':
-        intensities = pixels.min(axis=1)
-    elif intensity_type == 'sum' or 'total':
-        intensities = pixels.sum(axis=1)
+    if reduce_func == None:
```

Comment:
I vote for setting to `None` by default. BTW, the test would be `if reduce_func is None` instead of `if reduce_func == None`

Reply:
> Should I have default as np.mean, and leave out this check?

I'd say yes, except for the fact that this will not render nicely in the HTML documentation which will show that value as `reduce_func=<function mean at 0x7fc9686c9d40>` or something similar. So perhaps we should indeed use `None` as the default value. I'd definitely prefer @jni's suggestion with the try except block.

Reply:
Well, I kinda would like `reduce_func=None` to mean that the whole 2D array is returned... So I'm thinking that the doc problem should be a fix to sphinx rather than work around doing the wrong thing...

Reply:
Well, we could define a class that wraps the function:

```python
class wrapped_mean:

    def __call__(self, *args, **kwargs):
        return np.mean(*args, **kwargs)

    def __repr__(self):
        return "{}.{}".format(np.mean.__module__, np.mean.__qualname__)
```
and use `reduce_func=wrapped_mean()` as a default.

... just kidding. :smile: You're probably right. But we should definitely make sure to explain the default value in the parameter description then because the signature won't be helpful there.

Reply:
Perhaps something like this?

```python
    if reduce_func is None:
        pass
    else:    
        try:
            intensities = reduce_func(pixels, axis=1)
        except TypeError:  # function doesn't allow axis kwarg
            intensities = np.apply_along_axis(reduce_func, arr=pixels, axis=1)
```

or could give a message or warning to the user instead of `pass`. Example:
`print("To reduce the array along the linewidth axis to 1D, choose a reduce_func ufunc")`

This would output the intensities array as 2D for reduce_func=None, as @jni mentioned :

> I kinda would like reduce_func=None to mean that the whole 2D array is returned...

I'd reckon it's good that the user can have this array outputted without reduction.

Reply:
Am I wrong or `reduce_func` acts on linewidth "direction"? So following @jni interpretation of `reduce_func=None`, the returned "line" is 2D for gray scale images and 3D for RGB images.
If `linewidth == 1`, no need for reduction, so may be a warning can be raised only if `reduce_func is None and linewidth > 1`? What do you think?

Reply:
@rfezzani Correct. But I don't think we need a warning. The behavior for passing `None` should be clearly documented and we can assume that if the user explicitly passes in `None` the result is expected.

Also, passing in `reduce_func=lambda arr, axis: arr` will return the same as passing in `None`. But it's probably worth it to allow `None` for the users convenience.

Reply:
You're right, my phrasing was probably a tad off :). 

Just to note, if `reduce_func=None` and `linewidth=1`, the array has the shape (N,1), which is still 2D. 
However, if `reduce_func=np.mean`, for example, and `linewidth=1`, the array has the shape (N, ). Should we make the output consistent? Or should we leave it as is, because that is what it should output...?

Reply:
The above behaviour (None -> 2D, anything else -> 1D) is the consistent behaviour, even for line width = 1. I think the following is right:

```python
    if reduce_func is None:
        intensities = pixels
    else:    
        try:
            intensities = reduce_func(pixels, axis=1)
        except TypeError:  # function doesn't allow axis kwarg
            intensities = np.apply_along_axis(reduce_func, arr=pixels, axis=1)
```

Additionally, with regards to the representation in the docs, I think @lagru your suggestion is not so terrible or funny, but rather something that could be used everywhere! I'd use a functional approach though:

```python
def pretty_print(func):
    @functools.wraps
    def inner(*args, **kwargs):
        return func(*args, **kwargs)
    inner.__str__ = lambda self: func.__module__ + '.' + func.__name__
    inner.__repr__ = inner.__str__
    return inner
```

I think this should work? Then it would be `reduce_func=pretty_print(np.mean)`, and it would show up as `numpy.mean` in the docs.

Bonus points if you can get intersphinx to link to the documentation for numpy.mean :joy: 

**update:** Ok, I tried it, and it fails spectacularly:

```python
In [10]: mean = pretty_print(np.mean)                                                                                                                                                         

In [11]: str(mean)                                                                                                                                                                            
Out[11]: "functools.partial(<function update_wrapper at 0x7f531b8dd9d8>, wrapped=<function pretty_print.<locals>.inner at 0x7f531881d1e0>, assigned=('__module__', '__name__', '__qualname__', '__doc__', '__annotations__'), updated=('__dict__',))"
```

:joy: but anyway, all this to say, there's legs to this idea, and I think it'd be generally useful, so this decorator should live in `_shared` and eventually might be graduated to e.g. numpydoc itself.

Reply:
I don't really like to pass callable as default argument because the kind of problems that we are facing right now...
The behavior (None -> 2D, anything else -> 1D) being new, what about a deprecation cycle?
```python
if reduce_func is None:
    warn("profile_line offers now the ability to apply different reduction strategies "
         + "perpendicularly to profile direction. In future version default behavior will "
         + "be no reduction. To disable this warning, set reduce_func to numpy.mean",  
         FutureWarning, stacklevel=2)
    reduce_func = np.mean
try:
    intensities = reduce_func(pixels, axis=1)
except TypeError:  # function doesn't allow axis kwarg
    intensities = np.apply_along_axis(reduce_func, arr=pixels, axis=1)
```

Reply:
@rfezzani I think the mean should remain the default. I don't see the documentation issue as a big problem. Having None be the default and adding a deprecation message makes it more annoying to use in the most common case. Reducing the need for warnings is definitely a goal for scikit-image.

Reply:
I agree with @jni here. Default behavior should stay at `mean`. Accepting `None` as well is probably a good idea for less advanced users.

Also, your attempt probably didn't work because Python does method lookup on the class dictionary and not on the instance. So setting the `__repr__` attribute on the instance has no effect! However a class based wrapper should work fine. When I find the time I'll have a look into Sphinx's inspection for signatures and see if the problem should be fixed there or here.